java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -188,22 +206,83 @@ public BooleanConstraint shouldNotInverse() {
-    if (relationState.rejects(booleanConstraint)) {
-      return null;
+    return !relationState.rejects(booleanConstraint);
+  }
+
+  private List<SymbolicValue> transitiveRelations(ProgramState programState) {
+    BinaryRelation relation = binaryRelation();
+    return programState.getKnownRelations().stream()
+      .map(r -> r.deduceTransitiveOrSimplified(relation))
+      .filter(Objects::nonNull)
+      .map(RelationalSymbolicValue::binaryRelationToSymbolicValue)
+      .collect(Collectors.toList());
+  }
+
+  private static SymbolicValue binaryRelationToSymbolicValue(BinaryRelation binaryRelation) {
+    switch (binaryRelation.kind) {
+      case EQUAL:
+      case METHOD_EQUALS:
+      case LESS_THAN:
+        return new RelationalSymbolicValue(binaryRelation.kind, binaryRelation.leftOp, binaryRelation.rightOp);
+      case NOT_EQUAL:
+        return new NotSymbolicValue(new RelationalSymbolicValue(Kind.EQUAL, binaryRelation.leftOp, binaryRelation.rightOp));
+      case GREATER_THAN:
+        return new RelationalSymbolicValue(Kind.LESS_THAN, binaryRelation.rightOp, binaryRelation.leftOp);
+      case GREATER_THAN_OR_EQUAL:
+        return new NotSymbolicValue(new RelationalSymbolicValue(Kind.LESS_THAN, binaryRelation.leftOp, binaryRelation.rightOp));
+      case LESS_THAN_OR_EQUAL:
+        return new NotSymbolicValue(new RelationalSymbolicValue(Kind.LESS_THAN, binaryRelation.rightOp, binaryRelation.leftOp));
+      case NOT_METHOD_EQUALS:
+        return new NotSymbolicValue(new RelationalSymbolicValue(Kind.METHOD_EQUALS, binaryRelation.rightOp, binaryRelation.leftOp));
+      default:
+        throw new IllegalStateException("Unable to convert to relational SV " + binaryRelation);
     }
-    return programState;
   }
 
   @Override
   public BinaryRelation binaryRelation() {
-    return BinaryRelation.binaryRelation(kind, leftOp, rightOp);
+    if (binaryRelation == null) {
+      binaryRelation = BinaryRelation.binaryRelation(kind, leftOp, rightOp);
+    }
+    return binaryRelation;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    RelationalSymbolicValue that = (RelationalSymbolicValue) o;
+    return equalRelation(that);
+  }
+
+  private boolean equalRelation(RelationalSymbolicValue that) {
+    if (kind != that.kind) {
+      return false;
+    }
+    if (leftOp.equals(that.leftOp) && rightOp.equals(that.rightOp)) {
+      return true;
+    }
+    return isEquality() && leftOp.equals(that.rightOp) && rightOp.equals(that.leftOp);
+  }
+
+  private boolean isEquality() {
+    return kind == Kind.EQUAL || kind == Kind.METHOD_EQUALS;
+  }
+
+  @Override
+  public int hashCode() {
+    // hashCode doesn't depend on order of operands, to make commutative operators equal when operands are swapped
+    return kind.hashCode() + leftOp.hashCode() + rightOp.hashCode();
```

Comment:
wondering about collision here. I would have the instinct to multiply kind and sum of operand by two different prime number to make collision even less likely but I dunno how much of a superstition that is. Note that risk of collision is important given that hashcode will be used a lot as SV are keys of PMap in PS.